File: packages/python/spaces/starfish_spaces/inbox.py
"""Inbox client helpers.

Implements a monthly-sharded public-write ring buffer for sealed inter-user
messages (invites, resource grants, etc.).  Inbox paths follow the pattern
``/push/inbox/{identity}/{YYYY-MM}`` — each calendar month is a separate shard.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any, Optional, TypedDict

if TYPE_CHECKING:
    from starfish_sdk import StarfishClient
    from starfish_spaces.session import Session


class InboxElement(TypedDict):
    """A single element pulled from an inbox shard."""

    ts: int
    data: Any


def inbox_shard(now: Optional[datetime] = None) -> str:
    """Return the current UTC month shard string (``YYYY-MM``)."""
    dt = now or datetime.now(tz=timezone.utc)
    return dt.strftime("%Y-%m")


def inbox_shards(now: Optional[datetime] = None) -> list[str]:
    """Return the current and previous UTC month shard strings (``[current, prev]``)."""
    dt = now or datetime.now(tz=timezone.utc)
    current = dt.strftime("%Y-%m")
    # Previous month
    if dt.month == 1:
        prev = f"{dt.year - 1}-12"
    else:
        prev = f"{dt.year}-{dt.month - 1:02d}"
    return [current, prev]


async def pull_inbox(
    client: "StarfishClient",
    session: "Session",
    identity: str,
    shard: str,
    since: int = 0,
) -> list[InboxElement]:
    """Pull inbox items from ``inbox/{identity}/{shard}`` (best-effort, returns ``[]`` on error).

    Args:
        client:   An authenticated StarfishClient.
        session:  The active session (used for the layout).
        identity: The inbox owner's identity string (usually ``userId``).
        shard:    A month shard string (``YYYY-MM``) or ``"default"``.
        since:    Pull only elements with ``ts > since`` (0 = full pull).

    Returns:
        List of inbox elements (``{ts, data}``), oldest-first, or ``[]`` on error.
    """
    path = session.layout.inbox_pull(identity, shard)
    try:
        items = await client.pull(path, append_field="items", since=since, full=(since == 0))
        if isinstance(items, list):
            return [
                {"ts": el.get("ts", 0), "data": el.get("data")}
                for el in items
                if isinstance(el, dict)
            ]
        return []
    except Exception:
        return []


__all__ = [
    "InboxElement",
    "inbox_shard",
    "inbox_shards",
    "pull_inbox",
]

File: packages/ts/spaces/src/inbox.ts
/**
 * Inbox client helpers — read the `inbox/{identity}/{shard}` collection.
 *
 * Reads require a cap that covers `inbox/{identity}/**`. The owner's
 * `accountScope` / `linkedDeviceScope` already grants it — use `session.accountClient`.
 */
import type { StarfishClient } from "@drakkar.software/starfish-client"
import type { Session } from "./session.js"
import { makeAnonSpaceClient } from "./client.js"

/** Current UTC month shard in `YYYY-MM` format. */
export function inboxShard(): string {
  return new Date().toISOString().slice(0, 7)
}

/**
 * The current and previous UTC month shards — ensures an invite delivered near a
 * month boundary is still visible in the next month's scan.
 */
export function inboxShards(): string[] {
  const now = new Date()
  const y = now.getUTCFullYear()
  const m = now.getUTCMonth() // 0-indexed
  const current = `${y}-${String(m + 1).padStart(2, "0")}`
  const prevY = m === 0 ? y - 1 : y
  const prevM = m === 0 ? 12 : m
  const previous = `${prevY}-${String(prevM).padStart(2, "0")}`
  return [current, previous]
}

/** One inbox element as stored in the append-only shard. */
export interface InboxElement {
  ts: number
  data: Record<string, unknown>
}

/**
 * Append a pre-built element to `recipientUserId`'s inbox shard as an anonymous
 * public write, authored by the session identity.
 *
 * The inbox collection is `writeRoles:["public"]`, so no cap header is sent; the
 * Ed25519 author proof (the session's device keys) is bound to the document key and
 * surfaces the sender via `sealed.entry.addedBy` on the receive path. The caller is
 * responsible for sealing `element` first (see `resource-requests` for the canonical
 * seal-then-append pattern); use {@link inboxShard} for the current UTC shard.
 */
export async function appendToInbox(
  session: Session,
  recipientUserId: string,
  shard: string,
  element: Record<string, unknown>,
): Promise<void> {
  const anonClient = makeAnonSpaceClient({ baseUrl: session.baseUrl, namespace: session.namespace })
  await anonClient.appendAnonymous(
    session.layout.inboxPush(recipientUserId, shard),
    element,
    { edPubHex: session.keys.edPub, edPrivHex: session.keys.edPriv },
  )
}

/**
 * Pull one shard of `identity`'s inbox via `client`. Best-effort: returns `[]`
 * on any error (unreachable server, 403 stale cap, 404 empty month, etc.).
 */
export async function pullInbox(
  client: StarfishClient,
  identity: string,
  shard: string,
  session: Session,
): Promise<InboxElement[]> {
  return (await client
    .pull<InboxElement>(session.layout.inboxPull(identity, shard), {
      appendField: "items",
      full: true,
    })
    .catch(() => [])) as unknown as InboxElement[]
}
